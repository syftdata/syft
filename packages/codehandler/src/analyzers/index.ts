import { NamingCase, type SyftEventType } from '@syftdata/client';
import { convertCase } from '@syftdata/client/lib/utils';
import {
  type TypeField,
  type AST,
  type EventSchema,
  type Field
} from '@syftdata/common/lib/types';
import { type Project } from 'ts-morph';
import { AmplitudeAnalyser } from './plugins/amplitude';
import { GAAnalyser } from './plugins/gtag';
import { handleSourceFile } from './plugins/base';

export interface Usage {
  eventType: SyftEventType;
  eventName: string;
  typeFields: TypeField[];
}

function isEqualType(type1: TypeField, type2: TypeField): boolean {
  // TODO compare types.
  return true;
}
function getEventSchema(name: string, usageArray: Usage[]): EventSchema {
  const fieldNameToTypeMap = new Map<string, TypeField>();
  const fieldNameToCount = new Map<string, number>();
  usageArray.forEach((usage) => {
    usage.typeFields.forEach((field) => {
      const existingType = fieldNameToTypeMap.get(field.name);
      fieldNameToCount.set(
        field.name,
        (fieldNameToCount.get(field.name) ?? 0) + 1
      );
      if (existingType != null && !isEqualType(existingType, field)) {
        console.error(
          `Found incompatible field types for ${name}.${field.name} `
        );
      } else {
        fieldNameToTypeMap.set(field.name, field);
      }
    });
  });

  const fields: Field[] = [...fieldNameToTypeMap.entries()].map((entry) => {
    const fieldName = entry[0];
    const type = entry[1];
    const field: Field = {
      name: fieldName,
      type: {
        ...type.type,
        name: type.type.name,
        zodType: type.type.zodType
      },
      isOptional:
        type.isOptional ||
        (fieldNameToCount.get(fieldName) ?? 0) < usageArray.length
    };
    return field;
  });

  return {
    name: convertCase(name, NamingCase.PASCAL),
    eventType: usageArray[0].eventType,
    fields,
    zodType: ''
  };
}

export function analyzeAST(project: Project): AST {
  let usageDetails: Usage[] = [];
  const sourceFiles = project.getSourceFiles();
  const plugins = [new AmplitudeAnalyser(), new GAAnalyser()];
  sourceFiles.forEach((sourceFile) => {
    usageDetails = usageDetails.concat(handleSourceFile(sourceFile, plugins));
  });

  // merge usageDetails into arrays based on eventName field.
  const usageMapByEventName = new Map<string, Usage[]>();
  usageDetails.forEach((usage) => {
    const eventName = usage.eventName;
    if (usageMapByEventName.has(eventName)) {
      usageMapByEventName.get(eventName)?.push(usage);
    } else {
      usageMapByEventName.set(eventName, [usage]);
    }
  });
  const eventSchemas = Array.from(usageMapByEventName.entries()).map(
    ([eventName, usageDetails]) => getEventSchema(eventName, usageDetails)
  );

  return {
    eventSchemas,
    config: {
      projectName: 'Test',
      version: '0.0.1'
    }
  };
}
