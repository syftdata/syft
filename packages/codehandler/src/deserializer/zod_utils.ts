import { type TypeField } from '@syftdata/common/lib/types';

const TypeToZodMapping = {
  number: 'number()',
  string: 'string()',
  boolean: 'boolean()',
  Date: 'date()',
  Object: 'object()'
};

const SyftTypeToZodMapping = {
  Email: 'string().email()',
  Url: 'string().url()',
  CountryCode: 'string().length(2)',
  UUID: 'string().uuid()',
  CUID: 'string().cuid()',
  CUID2: 'string().cuid2()'
};

export const ANY_TYPE = 'any()';

export function getZodType(
  type: string,
  enumValues: string[],
  syfttype?: string
): string {
  if (enumValues.length > 1) {
    // For Enum, return in the format of enum(["Salmon", "Tuna", "Trout"])
    return `enum([${enumValues.join(', ')}])`;
  }

  if (syfttype !== undefined) {
    return SyftTypeToZodMapping[syfttype] ?? ANY_TYPE;
  }
  return TypeToZodMapping[type] ?? ANY_TYPE;
}

export const ZOD_ALLOWED_TAGS: Record<string, string[]> = {
  number: [
    'min',
    'max',
    'positive',
    'negative',
    'nonnegative',
    'nonpositive',
    'multipleOf',
    'finite'
  ],
  string: ['min', 'max', 'length', 'startsWith', 'endsWith']
};

export function getZodTypeForSchema(fields: TypeField[]): string {
  return `z.object({
    ${fields
      .map((field) => `\t${field.name}: ${field.type.zodType}`)
      .join(',\n')}
  })`;
}
