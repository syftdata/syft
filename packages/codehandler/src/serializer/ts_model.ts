import { type AST } from '@syftdata/common/lib/types';
import * as handlebars from 'handlebars/dist/cjs/handlebars';
import { EVENT_MODELS_TEMPLATE } from '../templates/model_template';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { isEmptyString } from '@syftdata/common/lib/utils';

// TODO: Memoize
function getEventModelsTemplate(): handlebars.HandlebarsTemplateDelegate<any> {
  return handlebars.compile(EVENT_MODELS_TEMPLATE, { noEscape: true });
}

export function serialize(ast: AST): string {
  const schemas = ast.eventSchemas.map((schema) => ({
    ...schema,
    fields: schema.fields.map((field) => ({
      ...field,
      has_documentation: !isEmptyString(field.documentation)
    })),
    eventType: SyftEventType[schema.eventType]
  }));
  return getEventModelsTemplate()({
    schemas
  });
}
