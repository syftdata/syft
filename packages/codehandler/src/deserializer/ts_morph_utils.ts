import {
  type JSDoc,
  type JSDocTag,
  type ts,
  type Type,
  type Symbol as TSSymbol,
  SyntaxKind
} from 'ts-morph';
import { logInfo } from '@syftdata/common/lib/utils';
import {
  type TypeField,
  type Field,
  type TypeSchema
} from '@syftdata/common/lib/types';
import { ANY_TYPE, getZodType, getZodTypeForSchema } from './zod_utils';

// const SyftypeModuleIndex = 'client/src/index".type.'
const SyftypeIndex = 'type.';

export function getTags(docs: JSDoc[]): Array<JSDocTag<ts.JSDocTag>> {
  return docs.reduce<Array<JSDocTag<ts.JSDocTag>>>((prev, curr) => {
    return [...prev, ...curr.getTags()];
  }, []);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function printType(type: Type): void {
  console.log('>>>>> printType ', type.getText());
  const proto = Object.getPrototypeOf(type);
  for (const member of Object.getOwnPropertyNames(proto)) {
    if (member.startsWith('is')) {
      if (typeof type[member] === 'function') {
        const result = type[member]();
        if (result === true) {
          console.log('>>> member ', member);
        }
      }
    }
  }
}

/**
 * This method returns TypeSchema of a field.
 * @param property
 * @returns
 */
function getTypeField(property: TSSymbol, debugName: string): TypeField {
  const declaration = property.getValueDeclarationOrThrow();
  const type = getTypeSchema(declaration.getType(), debugName);
  const hasQuestion =
    declaration.getFirstChildByKind(SyntaxKind.QuestionToken) != null;
  return {
    name: property.getName(),
    type,
    isOptional: hasQuestion
  };
}

/**
 * This method returns TypeSchema for complex objects like
 * class / interface / type.
 * @param typeObj
 * @returns
 */
function getTypeSchemaForComplexObject(
  typeObj: Type<ts.Type>,
  debugName: string
): TypeSchema {
  let name = typeObj.getSymbol()?.getEscapedName() ?? 'Unknown';
  if (typeObj.isAnonymous()) {
    name = typeObj.getAliasSymbol()?.getEscapedName() ?? name;
  }

  const typeFields = typeObj
    .getApparentProperties()
    .map((property) =>
      getTypeField(property, `${debugName}.${property.getEscapedName()}`)
    )
    .filter((field) => field != null) as Field[];

  const zodType = getZodTypeForSchema(typeFields);
  return {
    name,
    typeFields,
    zodType
  };
}

const KNOWN_TYPES = new Set(['Date', 'string', 'number', 'Object']);

/**
 * This method returns type-schema for all types of fields.
 * including primitives, enums, class/interface and types.
 * @param typeObj
 * @returns
 */
export function getTypeSchema(
  typeObj: Type<ts.Type>,
  debugName: string
): TypeSchema {
  let name = typeObj.getText();
  let enumValues: string[] = [];
  const isOptional = typeObj.isNullable();
  let syfttype: string | undefined;
  let foundUnsuppotedCloudType = false;
  if (typeObj.isClassOrInterface() || name.includes(SyftypeIndex)) {
    if (name.includes(SyftypeIndex)) {
      // TODO: all types that start with "type." are treated as syft-types.
      // Downside: if syft-types are imported as something else, this doesn't work.
      // if something else is imported as type. it gets aggected.
      const typeStrLoc = name.indexOf(SyftypeIndex);
      syfttype = name.slice(typeStrLoc + SyftypeIndex.length);
      name = 'string';
    } else if (!KNOWN_TYPES.has(name)) {
      foundUnsuppotedCloudType = true;
      return getTypeSchemaForComplexObject(typeObj, debugName);
    }
  } else if (typeObj.isObject()) {
    foundUnsuppotedCloudType = true;
    return getTypeSchemaForComplexObject(typeObj, debugName);
  } else if (typeObj.isUnion() && !typeObj.isBoolean()) {
    const subtypes = typeObj.getUnionTypes();
    let incompatibleSubType = false;
    enumValues = subtypes.map((subtype) => {
      if (subtype.isEnumLiteral()) {
        const enumVal = subtype.getLiteralValue()?.toString();
        if (enumVal != null) {
          return subtype.isStringLiteral() ? `"${enumVal}"` : enumVal;
        }
      } else if (
        subtype.isStringLiteral() ||
        subtype.isNumberLiteral() ||
        subtype.isBooleanLiteral()
      ) {
        return subtype.getText();
      }
      incompatibleSubType = true;
      return '';
    });
    if (!incompatibleSubType) {
      name = enumValues.join(' | ');
    } else {
      logInfo(
        `:warning: Unknown type is seen: "${debugName}: ${name}". using "any" for now`
      );
      name = 'any';
    }
  } else if (typeObj.isAny()) {
    // throw error.
    logInfo(`:warning: Any type is seen: "${debugName}".`);
    name = 'any';
    foundUnsuppotedCloudType = true;
  }

  if (foundUnsuppotedCloudType) {
    logInfo(
      `:prohibited: ${debugName} field is defined as ${name} type, which is not supported on the cloud.`
    );
  }

  let zodType = `z.${getZodType(name, enumValues, syfttype)}`;
  if (zodType === ANY_TYPE) {
    logInfo(
      `Used an unknown field type ${name}. No validations will be performed`
    );
  }

  if (isOptional) {
    zodType = `${zodType}.optional()`;
  }

  return {
    name,
    syfttype,
    zodType
  };
}
