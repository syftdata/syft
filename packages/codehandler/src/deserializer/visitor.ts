import {
  JSDocTypeTag,
  type ClassDeclaration,
  type Project,
  type PropertyDeclaration,
  type SourceFile,
  type Decorator,
  type JSDoc,
  SyntaxKind
} from 'ts-morph';
import {
  type EventSchema,
  type Field,
  type TypeField
} from '@syftdata/common/lib/types';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { logError, logVerbose } from '@syftdata/common/lib/utils';
import {
  extractKVPairsFromObjectLiteral,
  getTags,
  getTypeSchema
} from './ts_morph_utils';
import { getZodTypeForSchema, ZOD_ALLOWED_TAGS } from './zod_utils';
import {
  type DBEventSource,
  type DBFieldRelation
} from '@syftdata/common/lib/db_types';

interface EventProperties {
  eventType?: SyftEventType;
  dbSourceDetails?: DBEventSource;
}

interface FieldProperties {
  zodType?: string;
  dbRelation?: DBFieldRelation;
}

function getField(
  property: PropertyDeclaration,
  typeFields?: Map<string, TypeField>
): Field | undefined {
  const typeField = typeFields?.get(property.getName());
  if (typeField == null) {
    // TODO: add additional details
    logError(`Couldn't find type field ${property.getName()}`);
    return;
  }

  const fieldProps = extractFieldProperties(
    typeField,
    property.getDecorators(),
    property.getJsDocs()
  );
  typeField.type.zodType = fieldProps.zodType ?? typeField.type.zodType;

  return {
    ...typeField,
    dbRelation: fieldProps.dbRelation,
    documentation: property
      .getJsDocs()
      .map((doc) => doc.getInnerText())
      .join('\n'),
    defaultValue: property.getInitializer()?.getText()
  };
}

function extractFieldProperties(
  typeField: TypeField,
  decorators: Decorator[],
  docs: JSDoc[]
): FieldProperties {
  const tags = getTags(docs);
  let fieldRelationProperties: DBFieldRelation | undefined;
  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'relation') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@relation decorator requires an argument`);
        return;
      }
      const dbArg = args[0];
      if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
        const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
        if (
          kvPairs.has('table') &&
          kvPairs.has('references') &&
          kvPairs.has('fields')
        ) {
          fieldRelationProperties = {
            table: kvPairs.get('table') as string,
            references: kvPairs.get('references') as string[],
            fields: kvPairs.get('fields') as string[],
            isMany: typeField.type.isArray ?? false
          };
        }
      }
      if (fieldRelationProperties === undefined) {
        logError(
          `relation decorator requires table, references, fields properties`
        );
      }
    }
  });

  let zodType = typeField.type.zodType;
  const allowedZodTags = ZOD_ALLOWED_TAGS[typeField.name] ?? [];
  tags.forEach((tag) => {
    const name = tag.getTagName();
    if (allowedZodTags.includes(name)) {
      const paramString = tag.getCommentText()?.trim();
      if (paramString !== undefined && paramString.length > 0) {
        if (name === 'startsWith' || name === 'endsWith') {
          zodType = `${zodType}.${name}("${paramString}")`;
        } else {
          const paramInt = parseInt(paramString);
          zodType = `${zodType}.${name}(${!isNaN(paramInt) ? paramInt : ''})`;
        }
      }
    }
  });
  return {
    zodType,
    dbRelation: fieldRelationProperties
  };
}

function extractEventProperties(
  decorators: Decorator[],
  docs: JSDoc[]
): EventProperties {
  const tags = getTags(docs);
  let eventType: SyftEventType | undefined;
  let dbProperties: DBEventSource | undefined;

  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'eventtype') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@eventtype decorator requires arguments`);
        return;
      }
      const arg = args[0];
      const eventTypeString = arg.getText();
      if (!eventTypeString.startsWith('SyftEventType.')) {
        logError(`@eventtype decorator requires SyftEventType enum`);
        return;
      }
      eventType = SyftEventType[eventTypeString.replace('SyftEventType.', '')];
      if (eventType === SyftEventType.DB) {
        const dbArg = args[1];
        if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
          const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
          if (kvPairs.has('table') && kvPairs.has('on')) {
            dbProperties = {
              table: kvPairs.get('table') as string,
              on: kvPairs.get('on') as string,
              fields: kvPairs.get('fields') as string[]
            };
          }
        }
        if (dbProperties === undefined) {
          logError(
            `eventtype decorator with SyftEventType.DB requires table and on properties`
          );
        }
      }
    }
  });

  tags.forEach((tag) => {
    if (tag instanceof JSDocTypeTag || tag.getTagName() === 'type') {
      const fullTypeString = (tag as JSDocTypeTag)
        .getTypeExpression()
        ?.getTypeNode()
        ?.getText();
      if (fullTypeString === undefined) return;
      const typeString = fullTypeString.replace('SyftEventType.', '');
      if (typeString in SyftEventType) {
        eventType = SyftEventType[typeString as keyof typeof SyftEventType];
      }
    }
  });
  return { eventType, dbSourceDetails: dbProperties };
}

export function getEventSchema(
  project: Project,
  classObj: ClassDeclaration
): EventSchema {
  const name = classObj.getNameOrThrow();
  logVerbose(`Parsing event: ${name}`);
  const typeSchema = getTypeSchema(classObj.getType(), name);
  // for every field, get the TypeField object map.
  const typeFieldMap = typeSchema.typeFields?.reduce((prev, curr) => {
    prev.set(curr.name, curr);
    return prev;
  }, new Map<string, TypeField>());

  let fields = classObj
    .getProperties()
    .map((property) => {
      return getField(property, typeFieldMap);
    })
    .filter((field) => field != null) as Field[];
  const eventProps = extractEventProperties(
    classObj.getDecorators(),
    classObj.getJsDocs()
  );

  const fieldMap = fields.reduce((map, field) => {
    map.set(field.name, field);
    return map;
  }, new Map<string, Field>());

  // TODO: remove type field, which is the last one usually.
  let documentation = classObj
    .getJsDocs()
    .map((doc) => doc.getInnerText())
    .join('\n');
  const baseClass = classObj.getBaseClass();
  let baseSchema: EventSchema | undefined;
  if (baseClass !== null && baseClass !== undefined) {
    baseSchema = getEventSchema(project, baseClass); // eslint-disable-line prefer-const

    // merge base-class and class fields, de-dupe them.
    baseSchema.fields.forEach((field) => {
      if (!fieldMap.has(field.name)) {
        fieldMap.set(field.name, field);
      } else {
        if (fieldMap.get(field.name)?.type.name !== field.type.name) {
          logError(
            `${name}.${field.name} is an incompatible override compared to its base type.`
          );
        }
      }
    });

    if (documentation === '' && baseSchema.documentation != null) {
      documentation = baseSchema.documentation;
    }
    fields = Array.from(fieldMap.values());
  }

  const zodType = getZodTypeForSchema(fields);
  documentation = documentation
    .replace(/\n@type {SyftEventType.*}/g, '')
    .trim();
  return {
    name,
    eventType:
      eventProps.eventType ?? baseSchema?.eventType ?? SyftEventType.TRACK,
    exported: classObj.isExported(),
    documentation,
    fields,
    dbSourceDetails: eventProps.dbSourceDetails ?? baseSchema?.dbSourceDetails,
    zodType
  };
}

export function getEventSchemas(
  project: Project,
  source: SourceFile
): EventSchema[] {
  // Walk the tree to search for classes
  return source
    .getClasses()
    .map((classObj) => getEventSchema(project, classObj));
}

export function getImports(project: Project, source: SourceFile): string[] {
  return source.getImportDeclarations().map((dec) => {
    return dec.getText();
  });
}
