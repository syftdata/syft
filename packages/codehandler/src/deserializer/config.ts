import {
  type StaticConfig,
  SyftSinkType,
  SyftInputType
} from '@syftdata/common/lib/client_types';
import { type InputSource, type Sink } from '@syftdata/common/lib/types';
import type { ArrayLiteralExpression, Project } from 'ts-morph';
import { SyntaxKind, ObjectLiteralExpression } from 'ts-morph';
import { getConfigObject } from './ts_morph_utils';

const STATIC_CONFIG_TYPE = 'StaticConfig';

export function getConfigExpressionForProject(
  project: Project
): ObjectLiteralExpression | undefined {
  let syftConfig: ObjectLiteralExpression | undefined;
  for (const sourceFile of project.getSourceFiles()) {
    sourceFile.getVariableDeclarations().forEach((variable) => {
      let varialbeType = variable.getType().getText();
      const symbol = variable.getType().getSymbol();
      if (symbol !== undefined) {
        varialbeType = symbol.getEscapedName();
      }
      if (varialbeType !== STATIC_CONFIG_TYPE) return;
      syftConfig = variable.getInitializerIfKind(
        SyntaxKind.ObjectLiteralExpression
      );
    });
    if (syftConfig !== undefined) break;
  }
  return syftConfig;
}

export function getConfigFromExpression(
  expression: ObjectLiteralExpression
): StaticConfig {
  const val: StaticConfig = {
    projectName: 'test',
    version: '0.0.0'
  };
  const config = getConfigObject(expression);
  if (config === undefined) return val;
  return {
    ...val,
    ...config
  };
}

function getArrayLiteralExpression(
  project: Project,
  name: string
): ArrayLiteralExpression | undefined {
  let sinks: ArrayLiteralExpression | undefined;
  for (const sourceFile of project.getSourceFiles()) {
    const arrays = sourceFile
      .getVariableDeclarations()
      .map((variable) => {
        if (variable.getName() === name) {
          return variable.getInitializerIfKind(
            SyntaxKind.ArrayLiteralExpression
          );
        }
        return undefined;
      })
      .filter((x) => x !== undefined) as ArrayLiteralExpression[];
    if (arrays.length > 0) {
      sinks = arrays[0];
    }
    if (sinks !== undefined) break;
  }
  return sinks;
}

export function getSinks(project: Project): Sink[] {
  const sinks = getArrayLiteralExpression(project, 'sinks');
  if (sinks === undefined) return [];
  const sinkObjects = sinks
    .getElements()
    .map((element) => {
      if (element instanceof ObjectLiteralExpression) {
        return getConfigObject(element) as unknown as Sink;
      }
      return undefined;
    })
    .filter((x) => x !== undefined) as Sink[];
  return sinkObjects;
}

export function getInputs(project: Project): InputSource[] {
  const inputs = getArrayLiteralExpression(project, 'inputs');
  if (inputs === undefined) return [];
  const inputObjects = inputs
    .getElements()
    .map((element) => {
      if (element instanceof ObjectLiteralExpression) {
        return getConfigObject(element) as unknown as InputSource;
      }
      return undefined;
    })
    .filter((x) => x !== undefined) as InputSource[];
  return inputObjects;
}
