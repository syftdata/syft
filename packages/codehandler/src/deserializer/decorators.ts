import { JSDocTypeTag, type Decorator, type JSDoc, SyntaxKind } from 'ts-morph';
import { type TypeField } from '@syftdata/common/lib/types';
import { SyftEventType } from '@syftdata/common/lib/client_types';
import { logError } from '@syftdata/common/lib/utils';
import { extractKVPairsFromObjectLiteral, getTags } from './ts_morph_utils';
import { ZOD_ALLOWED_TAGS } from './zod_utils';
import {
  type DBEventSource,
  type DBFieldRelation
} from '@syftdata/common/lib/db_types';

export interface EventProperties {
  eventType?: SyftEventType;
  dbSourceDetails?: DBEventSource;
}

export interface FieldProperties {
  zodType?: string;
  dbRelation?: DBFieldRelation;
  rename?: string;
}

export function extractFieldProperties(
  typeField: TypeField,
  decorators: Decorator[],
  docs: JSDoc[]
): FieldProperties {
  const tags = getTags(docs);
  let fieldRelationProperties: DBFieldRelation | undefined;
  let rename: string | undefined;
  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'relation') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@relation decorator requires an argument`);
        return;
      }
      const dbArg = args[0];
      if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
        const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
        if (
          kvPairs.has('table') &&
          kvPairs.has('references') &&
          kvPairs.has('fields')
        ) {
          fieldRelationProperties = {
            table: kvPairs.get('table') as string,
            references: kvPairs.get('references') as string[],
            fields: kvPairs.get('fields') as string[],
            isMany: typeField.type.isArray ?? false
          };
        }
      }
      if (fieldRelationProperties === undefined) {
        logError(
          `relation decorator requires table, references, fields properties`
        );
      }
    } else if (name === 'rename') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@rename decorator requires an argument`);
        return;
      }
      const dbArg = args[0];
      if (dbArg.isKind(SyntaxKind.StringLiteral)) {
        rename = dbArg.getLiteralValue();
      }
    }
  });

  let zodType = typeField.type.zodType;
  const allowedZodTags = ZOD_ALLOWED_TAGS[typeField.name] ?? [];
  tags.forEach((tag) => {
    const name = tag.getTagName();
    if (allowedZodTags.includes(name)) {
      const paramString = tag.getCommentText()?.trim();
      if (paramString !== undefined && paramString.length > 0) {
        if (name === 'startsWith' || name === 'endsWith') {
          zodType = `${zodType}.${name}("${paramString}")`;
        } else {
          const paramInt = parseInt(paramString);
          zodType = `${zodType}.${name}(${!isNaN(paramInt) ? paramInt : ''})`;
        }
      }
    }
  });
  return {
    zodType,
    dbRelation: fieldRelationProperties,
    rename
  };
}

export function extractEventProperties(
  decorators: Decorator[],
  docs: JSDoc[]
): EventProperties {
  const tags = getTags(docs);
  let eventType: SyftEventType | undefined;
  let dbProperties: DBEventSource | undefined;

  decorators.forEach((decorator) => {
    const name = decorator.getName();
    if (name === 'eventtype') {
      const args = decorator.getArguments();
      if (args.length === 0) {
        logError(`@eventtype decorator requires arguments`);
        return;
      }
      const arg = args[0];
      const eventTypeString = arg.getText();
      if (!eventTypeString.startsWith('SyftEventType.')) {
        logError(`@eventtype decorator requires SyftEventType enum`);
        return;
      }
      eventType = SyftEventType[eventTypeString.replace('SyftEventType.', '')];
      if (eventType === SyftEventType.DB) {
        const dbArg = args[1];
        if (dbArg.isKind(SyntaxKind.ObjectLiteralExpression)) {
          const kvPairs = extractKVPairsFromObjectLiteral(dbArg);
          if (kvPairs.has('table') && kvPairs.has('on')) {
            dbProperties = {
              table: kvPairs.get('table') as string,
              on: kvPairs.get('on') as string,
              fields: kvPairs.get('fields') as string[]
            };
          }
        }
        if (dbProperties === undefined) {
          logError(
            `eventtype decorator with SyftEventType.DB requires table and on properties`
          );
        }
      }
    }
  });

  tags.forEach((tag) => {
    if (tag instanceof JSDocTypeTag || tag.getTagName() === 'type') {
      const fullTypeString = (tag as JSDocTypeTag)
        .getTypeExpression()
        ?.getTypeNode()
        ?.getText();
      if (fullTypeString === undefined) return;
      const typeString = fullTypeString.replace('SyftEventType.', '');
      if (typeString in SyftEventType) {
        eventType = SyftEventType[typeString as keyof typeof SyftEventType];
      }
    }
  });
  return { eventType, dbSourceDetails: dbProperties };
}
