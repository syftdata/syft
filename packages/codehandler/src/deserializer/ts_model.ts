import { Project, ScriptTarget } from 'ts-morph';
import { logDetail, logError, logFatal } from '@syftdata/common/lib/utils';
import {
  type InputSource,
  type AST,
  type Sink
} from '@syftdata/common/lib/types';
import { getEventSchemas } from './visitor';
import * as fs from 'fs';
import * as path from 'path';
import {
  getConfigExpressionForProject,
  getConfigFromExpression,
  getInputs,
  getSinks
} from './config';

// TODO: memoize this method.
export function createTSProject(filePaths: string[]): Project {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5, strictNullChecks: true }
  });

  // if directory, add glob style
  const files = filePaths.map((filePath) => {
    if (fs.existsSync(filePath) && fs.lstatSync(filePath).isDirectory()) {
      return path.join(filePath, '**', '!(*.d).ts');
    }
    return filePath;
  });
  if (files.length > 0) {
    project.addSourceFilesAtPaths(files);
  }
  return project;
}

export function readTSProject(
  sourceDir: string
): { project: Project; packageJson: any } | undefined {
  const tsConfigFilePath = `${sourceDir}/tsconfig.json`;
  if (!fs.existsSync(tsConfigFilePath)) {
    logError(
      `tsconfig.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read package.json file.
  const packageJsonPath = `${sourceDir}/package.json`;
  if (!fs.existsSync(packageJsonPath)) {
    logError(
      `package.json not found in ${sourceDir}. Make sure you are in the root of a typescript project.`
    );
    return;
  }
  // read all files in the project from source dir. and initialize ts-morph project
  const project = new Project({
    tsConfigFilePath
  });
  const packageJson = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
  return {
    project,
    packageJson
  };
}

export function generateASTForProject(project: Project): AST | undefined {
  const sourceFile = project.getSourceFile('events.ts');
  if (sourceFile == null) {
    logFatal(
      ':warning: events.ts file is not found. Please initialize your project using `syft init`'
    );
    return;
  }
  const relPath = path.relative('.', sourceFile.getFilePath());
  logDetail(`:heavy_check_mark: Parsing ${relPath}..`);
  const eventSchemas = getEventSchemas(project, sourceFile);

  const syftConfig = getConfigExpressionForProject(project);
  if (syftConfig === undefined) {
    logFatal(':warning: Config file is not found. Please provide one.');
    return;
  }
  const config = getConfigFromExpression(syftConfig);
  const sinks: Sink[] = getSinks(project);
  const inputs: InputSource[] = getInputs(project);
  return {
    eventSchemas,
    syftConfig,
    config,
    sinks,
    inputs
  };
}

export function deserialize(fileNames: string[]): AST | undefined {
  return generateASTForProject(createTSProject(fileNames));
}

export function deserializeFromCode(
  code: string,
  config: string
): AST | undefined {
  const project = new Project({
    compilerOptions: { target: ScriptTarget.ES5 }
  });
  project.createSourceFile(path.join('events.ts'), (writer) => {
    writer.write(code);
  });
  project.createSourceFile(path.join('config.ts'), (writer) => {
    writer.write(config);
  });
  return generateASTForProject(project);
}
