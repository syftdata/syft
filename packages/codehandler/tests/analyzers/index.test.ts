/* eslint-disable @typescript-eslint/no-unused-vars */
import { analyzeAST } from '../../src/analyzers/index';
import { createTSProject } from '../../src/deserializer/ts_model';

describe('analyze source code', () => {
  it('finds amplitude usages', async () => {
    const proj = createTSProject(['./tests/test_project/']);
    const ast = analyzeAST(proj);

    // expect(ast.eventSchemas).toHaveLength(4);
    expect(JSON.stringify(ast?.eventSchemas, null, 2)).toMatchSnapshot();
  });
});
