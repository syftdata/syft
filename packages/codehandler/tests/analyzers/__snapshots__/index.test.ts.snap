// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze source code finds amplitude usages 1`] = `
"[
  {
    "name": "EventAmpli1",
    "eventType": 0,
    "fields": [
      {
        "name": "argOpt",
        "type": {
          "name": "number",
          "zodType": "z.number().optional()",
          "isArray": false
        },
        "isOptional": true
      },
      {
        "name": "arg2",
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "isOptional": false
      }
    ],
    "zodType": ""
  },
  {
    "name": "EventAmpli2",
    "eventType": 0,
    "fields": [
      {
        "name": "eventName",
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "arg1",
        "type": {
          "name": "number",
          "zodType": "z.number()",
          "isArray": false
        },
        "isOptional": true
      },
      {
        "name": "arg2",
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "arg3",
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "isOptional": true
      }
    ],
    "zodType": ""
  },
  {
    "name": "EventAmpliPage",
    "eventType": 1,
    "fields": [
      {
        "name": "path",
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "arg2",
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "isOptional": false
      }
    ],
    "zodType": ""
  },
  {
    "name": "EventGa1",
    "eventType": 0,
    "fields": [
      {
        "name": "arg1",
        "type": {
          "name": "number",
          "zodType": "z.number()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "arg2",
        "type": {
          "name": "boolean",
          "zodType": "z.boolean()",
          "isArray": false
        },
        "isOptional": false
      }
    ],
    "zodType": ""
  },
  {
    "name": "UserIdentify",
    "eventType": 3,
    "fields": [
      {
        "name": "userid",
        "type": {
          "name": "number",
          "zodType": "z.number()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "email",
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "isOptional": false
      },
      {
        "name": "user",
        "type": {
          "name": "string",
          "zodType": "z.string()",
          "isArray": false
        },
        "isOptional": false
      }
    ],
    "zodType": ""
  }
]"
`;
